### PR TITLE
simple view changer

### PR DESCRIPTION
**Pull request by HagarMeir**
Signed-off-by: Hagar Meir <hagar.meir@ibm.com>

### CODE REVIEW

**Comment by HagarMeir** [comment]
Issue: https://github.com/SmartBFT-Go/consensus/issues/178

File: internal/bft/viewchanger.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+				Signer:      v.SelfID,
+				Signature:   sig,
+			},
+		},
+	}
+	return msg
+}
+
+func (v *ViewChanger) processViewDataMsg() {
+	for viewNum, votes := range v.viewDataMsgs {
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: can you extract the loop body into a method?

File: internal/bft/viewchanger.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+		if vc.NextView != v.CurrView+1 { // accept view change only to immediate next view number
+			v.Logger.Warnf("%d got viewChange message %v from %d with view %d, expected view %d", v.SelfID, m, sender, vc.NextView, v.CurrView+1)
+			return
+		}
+		v.viewChangeMsgs.registerVote(sender, m)
+		return
+	}
+
+	//viewData message
+	if vd := m.GetViewData(); vd != nil {
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: can you please extract this block to a method?

File: internal/bft/viewchanger.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+		if vd.Signer != sender {
+			v.Logger.Warnf("%d got viewData message %v from %d, but signer %d is not the sender %d", v.SelfID, m, sender, vd.Signer, sender)
+			return
+		}
+		if err := v.Verifier.VerifySignature(types.Signature{Id: vd.Signer, Value: vd.Signature, Msg: vd.RawViewData}); err != nil {
+			v.Logger.Warnf("%d got viewData message %v from %d, but signature is invalid, err: ", v.SelfID, m, sender, err)
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: %v missing after "err:"

File: internal/bft/viewchanger.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+			return
+		}
+		rvd := &protos.ViewData{}
+		if err := proto.Unmarshal(vd.RawViewData, rvd); err != nil {
+			v.Logger.Errorf("%d was unable to unmarshal viewData message from %d, error: ", v.SelfID, sender, err)
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: %v missing after "error:"

File: internal/bft/viewchanger.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+				NextView: v.NextView,
+				Reason:   "", // TODO add reason
+			},
+		},
+	}
+	v.Comm.BroadcastConsensus(msg) // TODO periodically send msg
+}
+
+func (v *ViewChanger) processViewChangeMsg(vote *vote) {
+	if uint64(len(v.viewChangeMsgs.voted)) > v.F { // join view change
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: this makes you send a "view change" every time you get a "view change" from someone.
Better change this to **== v.F+1** instead.

File: internal/bft/viewchanger.go (hunk 2 above)
Comment (yacovm) [inline internal/bft/viewchanger.go]: below this line, there are lots of un-tested cases.

File: internal/bft/viewchanger.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+func (v *ViewChanger) processViewDataMsg() {
+	for viewNum, votes := range v.viewDataMsgs {
+		if viewNum < v.CurrView { // clean old votes
+			delete(v.viewDataMsgs, viewNum)
+			continue
+		}
+		num := len(votes.voted)
+		if num >= v.Quorum { // need enough (quorum) data to continue
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: better yet - check the opposite condition **(num < v.Quorum)** and **continue**

File: internal/bft/viewchanger.go (hunk 2 above)
Comment (HagarMeir) [reply internal/bft/viewchanger.go]: I still have a lot of tests to write...

File: internal/bft/viewchanger.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+			// TODO handle data
+			repeated := make([]*protos.SignedViewData, num)
+			n := 0
+			for n < num {
+				v := <-votes.votes
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: you can close the **votes.votes** and then just foreach on the closed channel

File: internal/bft/viewchanger.go [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+					NewView: &protos.NewView{
+						SignedViewData: repeated,
+					},
+				},
+			}
+			v.Comm.BroadcastConsensus(msg)
+			v.HandleMessage(v.SelfID, msg) // also send to myself
+
+			delete(v.viewDataMsgs, viewNum) // don't need them anymore
+			break
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: return instead of break?

File: internal/bft/viewchanger.go [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+	nodesMap := make(map[uint64]struct{}, v.N)
+	valid := 0
+	for _, svd := range repeated {
+		if _, exist := nodesMap[svd.Signer]; exist {
+			continue // seen data from this node already
+		}
+		nodesMap[svd.Signer] = struct{}{}
+
+		if err := v.Verifier.VerifySignature(types.Signature{Id: svd.Signer, Value: svd.Signature, Msg: svd.RawViewData}); err != nil {
+			continue // invalid signature
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: please log the error in warningf

File: internal/bft/viewchanger.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+		}
+
+		vd := &protos.ViewData{}
+		if err := proto.Unmarshal(svd.RawViewData, vd); err != nil {
+			v.Logger.Errorf("%d was unable to unmarshal a viewData from the newView message, error: ", v.SelfID, err)
+			continue
+		}
+
+		if vd.NextView != v.CurrView {
```
Comment (yacovm) [inline internal/bft/viewchanger.go]: please log an error as the leader shouldn't send a message that isn't useful to us.

File: internal/bft/viewchanger_test.go [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+	}
+
+	vc.Start()
+
+	vd := &protos.ViewData{
+		NextView: 1,
+	}
+	vdBytes := bft.MarshalOrPanic(vd)
+	msg := &protos.Message{
+		Content: &protos.Message_ViewData{
```
Comment (yacovm) [inline internal/bft/viewchanger_test.go]: we send a view data without receiving a view change beforehand... i think the production code should expect to see the view change before digesting view data messages.

File: internal/bft/viewchanger_test.go [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+	vc.Start()
+
+	// create a valid viewData message
+	vd := &protos.ViewData{
+		NextView: 2,
+	}
+	vdBytes := bft.MarshalOrPanic(vd)
+	repeated := make([]*protos.SignedViewData, 0)
+	var node uint64
+	for node < uint64(vc.Quorum) {
```
Comment (yacovm) [inline internal/bft/viewchanger_test.go]: instead of node you can check the size of repeated.

File: internal/bft/viewchanger.go (hunk 6 above)
Comment (HagarMeir) [reply internal/bft/viewchanger.go]: not in for loop anymore so ignoring this

File: internal/bft/viewchanger.go (hunk 8 above)
Comment (HagarMeir) [reply internal/bft/viewchanger.go]: not in for loop anymore so ignoring this

File: internal/bft/viewchanger_test.go (hunk 11 above)
Comment (HagarMeir) [reply internal/bft/viewchanger_test.go]: Initialized viewChanger as if viewChange process was done

**Pull request closed without merging**